### PR TITLE
nixos-icons: recolor white logo

### LINKED ISSUE #292: gnome: NixOS logo on GDM login screen is always white
I recall this is part of a derivation which holds various NixOS logos, which we can override to use the foreground color.

### PR DESCRIPTION
**Pull request by danth**
Should we theme the colored logo too? I'm not sure, since it could look nice, but also the colors represent NixOS, and we don't usually change the color of other brands.

Fixes #292

### CODE REVIEW

File: modules/nixos/nixos.nix [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+  options.stylix.targets.nixos.enable =
+    config.lib.stylix.mkEnableTarget "the NixOS logo" true;
+
+  config.nixpkgs.overlays = lib.mkIf config.stylix.targets.nixos.enable [(self: super: {
+    nixos-icons = super.nixos-icons.overrideAttrs (oldAttrs: {
+      patchPhase = ''
+        substituteInPlace logo/white.svg \
+          --replace-fail '#ffffff' '#${config.lib.stylix.colors.base05}'
+        ${lib.getExe pkgs.resvg} logo/white.svg logo/white.png
+      '';
```
Comment (trueNAHO) [inline modules/nixos/nixos.nix:13]: The logo is licensed under the CC-BY 4.0 license:

> License
> -------
>
> The logo is made available under a [CC-BY license](https://creativecommons.org/licenses/by/4.0/).
>
>  * [Nix-dev mailing list, Sun Oct 2 08:49:55 CEST 2016](https://nixos.org/nix-dev/2016-October/021876.html)
>
> (Source: https://github.com/NixOS/nixos-artwork/blob/35ebbbf01c3119005ed180726c388a01d4d1100c/logo/README.md#license)

As such, we are free to adapt, and therefore recolor, the original NixOS logo under the following terms:

> # Under the following terms:
>
> **Attribution** — You must give [appropriate credit](https://creativecommons.org/licenses/by/4.0/#ref-appropriate-credit), provide a link to the license, and [indicate if changes were made](https://creativecommons.org/licenses/by/4.0/#ref-indicate-changes). You may do so in any reasonable manner, but not in any way that suggests the licensor endorses you or your use.
>
> **No additional restrictions** — You may not apply legal terms or technological measures that legally restrict others from doing anything the license permits.
>
> (Source: https://creativecommons.org/licenses/by/4.0/)

To adhere to the *attribution* terms, we should extend the `patchPhase` such that it adds a comment to the file. Note that the original NixOS logo URL should be automatically generated to avoid wrong credit. Maybe this could be done by accessing the `flake.lock` version:

```xml
<!-- The original NixOS logo (https://github.com/NixOS/nixos-artwork/tree/35ebbbf01c3119005ed180726c388a01d4d1100c/logo) has been recolored to #<COLOR>. -->
```

Considering that [Stylix is licensed under MIT](https://github.com/danth/stylix/blob/2221c7d61b2e10b17df6c6795b4678fb59a0a92a/LICENSE), we adhere to the *No additional restrictions* terms.

Note: I am not a lawyer, and the information provided here should not be considered as legal advice.

**Review by trueNAHO — CHANGES_REQUESTED** [review]

File: modules/nixos/nixos.nix (hunk 1 above)
Comment (danth) [reply modules/nixos/nixos.nix]: I added the required attribution.

Below is an example of the comment, which varies based on the source URL and scheme name.

```svg
<!-- The original NixOS logo from https://github.com/NixOS/nixos-artwork/archive/488c22aad523c709c44169d3e88d34b4691c20dc.tar.gz modified to use colors from Catppuccin Macchiato. This file is licensed under https://creativecommons.org/licenses/by/4.0 -->
```

File: modules/nixos/nixos.nix [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+with config.lib.stylix.colors;
+
+{
+  options.stylix.targets.nixos.enable =
+    config.lib.stylix.mkEnableTarget "the NixOS logo" true;
+
+  config.nixpkgs.overlays = lib.mkIf config.stylix.targets.nixos.enable [(self: super: {
+    nixos-icons = super.nixos-icons.overrideAttrs (oldAttrs: {
+      patchPhase = ''
+        substituteInPlace logo/white.svg --replace-fail '#ffffff' '#${base05}'
```
Comment (trueNAHO) [inline modules/nixos/nixos.nix:12]: Nitpick: add empty line to improve readability:

```suggestion
        substituteInPlace logo/white.svg --replace-fail '#ffffff' '#${base05}'
        
```

File: modules/nixos/nixos.nix [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+        sed -i '2i<!-- The original NixOS logo from ${oldAttrs.src.url} modified to use colors from ${scheme}. This file is licensed under https://creativecommons.org/licenses/by/4.0 -->' logo/white.svg
```
Comment (trueNAHO) [inline modules/nixos/nixos.nix:13]: Improve readability by wrapping:

```suggestion
        sed \
          --in-place \
          '2i<!-- The original NixOS logo from ${oldAttrs.src.url} modified to use colors from ${scheme}. This file is licensed under https://creativecommons.org/licenses/by/4.0 -->' \
          logo/white.svg
```

We should explain why the comment cannot be added to the first line:

```suggestion
        # Add the XML comment to the second line to avoid conflicts with the XML
        # declaration on the first line.
        sed \
          --in-place \
          '2i<!-- The original NixOS logo (${oldAttrs.src.url}) is licensed under https://creativecommons.org/licenses/by/4.0 and has been modified to match the '${scheme}' color scheme. -->' \
          logo/white.svg
```

I would rephrase the comment as follows:

```suggestion
        # Add the XML comment to the second line to avoid conflicts with the XML
        # declaration on the first line.
        sed \
          --in-place \
          '2i<!-- The original NixOS logo (${oldAttrs.src.url}) is licensed under https://creativecommons.org/licenses/by/4.0 and has been modified to match the '${scheme}' color scheme. -->' \
          logo/white.svg
```

**Review by trueNAHO — CHANGES_REQUESTED** [review]

File: modules/nixos/nixos.nix (hunk 3 above)
Comment (danth) [reply modules/nixos/nixos.nix:13]: Note that the URL points to the repository as a whole so we should still include the word "from" (or something to that effect)

Comment (trueNAHO) [reply modules/nixos/nixos.nix:13]: > Note that the URL points to the repository as a whole so we should still include the word "from" (or something to that effect)

Agreed.

Howerver, if `logo/white.svg` stems from `oldAttrs.src.url`, then we could specify the exact file path as `<GITHUB_REPO>/tree/<REV>/logo/white.svg`. The hard-coded file URL should remain correct as long as the commands relying on `logo/white.svg` are still working.

Comment (danth) [reply modules/nixos/nixos.nix]: `oldAttrs.src` only tells us the Git URL `https://github.com/NixOS/nixos-artwork.git` or the download URL `https://github.com/NixOS/nixos-artwork/archive/488c22aad523c709c44169d3e88d34b4691c20dc.tar.gz`. By trying to construct a more specific URL, we would be relying on the source being hosted on GitHub and not some other provider.

**Review by trueNAHO — APPROVED** [review]

**Comment by danth** [comment]
Before:


After:

**Comment by danth** [comment]
One last thing: I'm not sure whether naming the target `nixos` is confusing.

**Comment by trueNAHO** [comment]
> One last thing: I'm not sure whether naming the target `nixos` is confusing.

Good point.

`logo` would be too generic and might conflict with https://github.com/danth/stylix/issues/242. `logo-nixos` or `nixos-logo` seem more appropriate.

Here is my take on both variants:

- `logo-nixos`
    - This implies a `nixos` variant of the (or a future) `logo` module.

- `nixos-logo`
    - This is the logo of NixOS.

I would pick `nixos-logo`.

**Comment by danth** [comment]
Perhaps we should match the package name and use `nixos-icons`.

**Comment by trueNAHO** [comment]
> Perhaps we should match the package name and use `nixos-icons`.

Sounds good to me.
